CRITEO - Use eviction to delete pod during roll restart

geobeau opened the pull request:
The roll restart code is prone to race conditions:
Calling the delete API is fast as it doesn't wait for actual termination
of the pod. The reconciler is called back almost instantly to handle
the other pools. The STS is not yet reporting the change of readiness
of one of the replica, and elasticsearch is not reporting the status as
yellow/red because the pod is not dead yet. So the operator thinks
it's okay to kill another pod.
Fixing the race condition can be done by introducing more checks:

- Using the `CountRunningPodsForNodePool` function, that is
more reliable than checking the status of sts. The status being
eventually consistent. This function list directly the pods.
- Use the Eviction API as it uses Compare And Swap
to ensure consistency with the budget. It can work without PDB
but it might not protect as efficiently against race conditions.

Also added logs to help debug race conditions in the future.

pgoron commented on opensearch-operator/pkg/reconcilers/rollingRestart.go:
```diff
@@ -190,12 +191,14 @@ func (r *RollingRestartReconciler) restartStatefulSetPod(sts *appsv1.StatefulSet
 	}
 
-	err = r.Delete(r.ctx, &corev1.Pod{
+	eviction := &policyv1.Eviction{}
```
suggesting to keep upstream code commented (with /* */ block to avoid as much as possible conflicts when bumping operator), and to re-explain here rational of our change, ie using eviction instead of Delete to respect PDB during roll-restart and with advantage of fixing concurrent roll-restart on node pools. A reference to upstream issue would be nice too.

geobeau replied:
Done

geobeau replied:
Undone !

pgoron commented on opensearch-operator/pkg/reconcilers/rollingRestart.go:
```diff
@@ -78,11 +80,27 @@ func (r *RollingRestartReconciler) Reconcile() (ctrl.Result, error) {
-					Requeue:      true,
-					RequeueAfter: 10 * time.Second,
-				}, nil
+			// if sts.Status.ReadyReplicas != pointer.Int32Deref(sts.Spec.Replicas, 1) {
```
wrap commented upstream code into block comments /* */ (less risk of conflicts during bump)

geobeau committed:
CRITEO - Use eviction to delete pod during roll restart

The roll restart code is prone to race conditions:
Calling the delete API is fast as it doesn't wait for actual termination
of the pod. The reconciler is called back almost instantly to handle
the other pools. The STS is not yet reporting the change of readiness
of one of the replica, and elasticsearch is not reporting the status as
yellow/red because the pod is not dead yet. So the operator thinks
it's okay to kill another pod.
Fixing the race condition can be done by introducing more checks:

- Using the `CountRunningPodsForNodePool` function, that is
more reliable than checking the status of sts. The status being
eventually consistent. This function list directly the pods.
- Use the Eviction API as it uses Compare And Swap
to ensure consistency with the budget. It can work without PDB
but it might not protect as efficiently against race conditions.

Also added logs to help debug race conditions in the future.
